File: Top 100 Codes/03-Codes for Recursion/12-Last non-zero digit in factorial/Lastnonzerodigitinfactorial.py
def last_non_zero_digit(n):
    result = 1
    for i in range(1, n + 1):
        result *= i

        # Remove trailing zeros by dividing by 10
        while result % 10 == 0:
            result //= 10

        # Keep only the last 5 digits to prevent overflow
        result %= 100000

    # Return the last non-zero digit
    return result % 10

# Input
n = int(input("Enter a number: "))
last_digit = last_non_zero_digit(n)
print(f"The last non-zero digit in {n}! is: {last_digit}")

File: Top 100 Codes/03-Codes for Recursion/12-Last non-zero digit in factorial/Lastnonzerodigitinfactorial.cpp
#include <iostream>
using namespace std;

// Function to find the last non-zero digit of factorial
int lastNonZeroDigit(int n) {
    int result = 1;
    
    for (int i = 1; i <= n; i++) {
        result *= i;

        // Remove trailing zeros by dividing by 10
        while (result % 10 == 0) {
            result /= 10;
        }

        // Keep only the last 5 digits to prevent overflow
        result %= 100000;
    }

    // Return the last non-zero digit
    return result % 10;
}

int main() {
    int n;
    cout << "Enter a number: ";
    cin >> n;

    int lastDigit = lastNonZeroDigit(n);
    cout << "The last non-zero digit in " << n << "! is: " << lastDigit << endl;

    return 0;
}
